packages/flutter/lib/src/material/scaffold.dart

```diff
@@ -37,10 +40,208 @@ enum _ScaffoldSlot {
+
+  @override
+  Animation<Offset> getOffsetAnimation({Offset begin, Offset end, Animation<double> parent}) {
+    return new Tween<Offset>(begin: begin, end: end).chain(new CurveTween(curve: const Threshold(0.5))).animate(parent);
+  }
+
+  @override
+  Animation<double> getScaleAnimation({Animation<double> parent}) {
+    final Curve curve = const Interval(0.5, 1.0, curve: Curves.ease);
+    return new _MaxAnimation<double>(
+      new ReverseAnimation(new CurveTween(curve: curve.flipped).animate(parent)),
+      new CurveTween(curve: curve).animate(parent),
+    );
+  }
+
+  @override
+  Animation<double> getRotationAnimation({Animation<double> parent}) {
+    // Because we only see the last half of the rotation tween, 
+    // it needs to go twice as far.
+    final Tween<double> rotationTween = new Tween<double>(
+      begin: 1.0 - _kFloatingActionButtonTurnInterval * 2, 
+      end: 1.0,
+    );
+    // This rotation will turn on the way in, but not on the way out.
+    return new _MaxAnimation<double>(
+      rotationTween.animate(parent),
+      new ReverseAnimation(new CurveTween(curve: const Threshold(0.5)).animate(parent)),
+    );
+  }
+}
+
+class _MaxAnimation<T> extends CompoundAnimation<T> {
+  _MaxAnimation(Animation<T> first, Animation<T> next): super(first: first, next: next);
+
+  @override
+  T get value => math.max(first.value, next.value);
+}
+
+/// The geometry of the [Scaffold] before it finishes laying out.
+/// 
+/// The Scaffold passes this prelayout geometry to its [FabPositioner].
+/// 
+/// For a description of the Scaffold's geometry after it has
+/// finished laying out, see the [ScaffoldGeometry].
+@immutable
+class ScaffoldPrelayoutGeometry {
+  const ScaffoldPrelayoutGeometry({this.bottomSheetSize, this.contentBottom, this.contentTop, this.fabSize, this.horizontalFabPadding, this.scaffoldSize, this.snackBarSize, this.textDirection});
+
+  /// The [Size] of the [Scaffold]'s [FloatingActionButton] (if available).
+  /// 
+  /// The Scaffold will determine the fabSize if 
+  /// `Scaffold.floatingActionButton` is not null.
+  final Size fabSize;
+
+  /// The [Size] of the [Scaffold]'s [BottomSheet] (if available).
+  /// 
+  /// The Scaffold will determine the bottomSheetSize if the Scaffold
+  /// is currently showing a [BottomSheet].
+  final Size bottomSheetSize;
+
+  /// The height from the [Scaffold]'s top where its body ends.
+  final double contentBottom;
+
+  /// The height from the [Scaffold]'s top where its body starts.
+  final double contentTop;
+
+  /// The minimum horizontal padding the [FloatingActionButton] should observe.
+  final double horizontalFabPadding;
+
+  /// The [Size] of the whole [Scaffold].
+  final Size scaffoldSize;
+
+  /// The [Size] of the [Scaffold]'s [SnackBar] (if available).
+  /// 
+  /// The Scaffold will determine the snackBarSize if the Scaffold
+  /// is currently showing a [SnackBar].
+  final Size snackBarSize;
+
+  /// The [Scaffold]'s [TextDirection].
```

Comment:
Scaffold doesn't have a TextDirection, so this probably needs more detail to describe what it really is...

Reply:
I've clarified that it's the TextDirection of the Scaffold's BuildContext.